core/src/main/java/edu/wpi/grip/core/http/GripServer.java

```diff
@@ -0,0 +1,404 @@
+        }
+    }
+    
+    
+    /**
+     * Restarts the server on the current port.
+     */
+    public void restart() {
+        try {
+            stop();
+            server = HttpServer.create(new InetSocketAddress("localhost", getPort()), BACKLOG);
+            start();
+        } catch (IOException | IllegalStateException ex) {
+            Logger.getLogger(GripServer.class.getName()).log(Level.SEVERE, null, ex);
+            throw new GripException("Could not restart GripServer", ex);
+        }
+    }
+    
+    /**
+     * Gets the port this server is running on.
+     */
+    public int getPort() {
+        return server.getAddress().getPort();
+    }
+
+    /**
+     * Parses request parameters from a URI. For example, the URI
+     * "http://roborio-190-frc.local:8080/GRIP/data?foo=1;bar=2" will be parsed to
+     * {"foo"="1", "bar"="2"}.
+     */
+    private static Map<String, List<String>> getRequestParameters(final URI requestUri) {
+        final Map<String, List<String>> requestParameters = new LinkedHashMap<>();
+        final String requestQuery = requestUri.getRawQuery();
+        if (requestQuery != null) {
+            final String[] rawRequestParameters = requestQuery.split("[&;]"); // [foo=1, bar=2, ...]
+            for (final String rawRequestParameter : rawRequestParameters) {
+                final String[] requestParameter = rawRequestParameter.split("=", 2); // [foo, 1]
+                final String requestParameterName = decodeUrlComponent(requestParameter[0]);
+                requestParameters.putIfAbsent(requestParameterName, new ArrayList<>());
+                final String requestParameterValue = requestParameter.length > 1 ? decodeUrlComponent(requestParameter[1]) : null;
+                requestParameters.get(requestParameterName).add(requestParameterValue);
+            }
+        }
+        return requestParameters;
+    }
+
+    private static String decodeUrlComponent(final String urlComponent) {
+        try {
+            return URLDecoder.decode(urlComponent, CHARSET.name());
+        } catch (final UnsupportedEncodingException ex) {
+            // Will never happen, UTF-8 is supported by all JVMs
+            throw new GripException(ex);
+        }
+    }
+
+    /**
+     * Reads all incoming bytes from an HTTP connection.
+     */
+    private static byte[] readBytes(final HttpExchange he) throws IOException {
+        final InputStream in = he.getRequestBody();
+        ByteArrayOutputStream b = new ByteArrayOutputStream();
+        byte[] buf = new byte[4096];
+        for (int n = in.read(buf); n > 0; n = in.read(buf)) {
+            b.write(buf, 0, n);
+        }
+        return b.toByteArray();
+    }
+
+    @Subscribe
+    public void settingsChanged(ProjectSettingsChangedEvent event) {
+        int port = event.getProjectSettings().getServerPort();
+        if (port != getPort()) {
+            stop();
+            server = serverFactory.create(port);
+            start();
```

Comment:
What if the server was stopped when the project settings changed? Now you've just restarted the server because the settings have changed.